jerry-main/main-unix.c

```diff
@@ -629,99 +640,136 @@ main (int argc,
 
-          receive_status = jerry_debugger_wait_for_client_source (wait_for_source_callback,
-                                                                  NULL,
-                                                                  &run_result);
-
-          if (receive_status == JERRY_DEBUGGER_SOURCE_RECEIVE_FAILED)
+          do
           {
-            ret_value = jerry_create_error (JERRY_ERROR_COMMON,
-                                            (jerry_char_t *) "Connection aborted before source arrived.");
+            jerry_value_t run_result;
+
+            receive_status = jerry_debugger_wait_for_client_source (wait_for_source_callback,
+                                                                    NULL,
+                                                                    &run_result);
+
+            if (receive_status == JERRY_DEBUGGER_SOURCE_RECEIVE_FAILED)
+            {
+              ret_value = jerry_create_error (JERRY_ERROR_COMMON,
+                                              (jerry_char_t *) "Connection aborted before source arrived.");
+            }
+
+            if (receive_status == JERRY_DEBUGGER_SOURCE_END)
+            {
+              jerry_port_log (JERRY_LOG_LEVEL_DEBUG, "No more client source.\n");
+            }
+
+            if (jerry_value_is_abort (run_result))
+            {
+              ret_value = jerry_acquire_value (run_result);
+            }
+
+            jerry_release_value (run_result);
           }
+          while (receive_status == JERRY_DEBUGGER_SOURCE_RECEIVED);
 
-          if (receive_status == JERRY_DEBUGGER_SOURCE_END)
+          if (receive_status != JERRY_DEBUGGER_CONTEXT_RESET_RECEIVED)
           {
-            jerry_port_log (JERRY_LOG_LEVEL_DEBUG, "No more client source.\n");
+            break;
           }
 
-          jerry_release_value (run_result);
-        }
-        while (receive_status == JERRY_DEBUGGER_SOURCE_RECEIVED);
+          init_engine (flags, true, debug_port);
 
-        if (receive_status != JERRY_DEBUGGER_CONTEXT_RESET_RECEIVED)
-        {
-          break;
+          ret_value = jerry_create_undefined ();
         }
+      }
 
-        jerry_cleanup ();
+    }
 
-        jerry_init (flags);
-        jerry_debugger_init (debug_port);
+    bool restart = false;
 
-        register_js_function ("assert", jerryx_handler_assert);
-        register_js_function ("gc", jerryx_handler_gc);
-        register_js_function ("print", jerryx_handler_print);
+    if (jerry_is_feature_enabled (JERRY_FEATURE_DEBUGGER) && jerry_value_is_abort (ret_value))
+    {
+      jerry_value_t abort_value = jerry_get_value_from_error (ret_value, false);
+      if (jerry_value_is_string (abort_value))
+      {
+        jerry_char_t str_buf[6];
+        jerry_value_t str_val = jerry_value_to_string (abort_value);
+        jerry_size_t str_size = jerry_get_string_size (str_val);
+        jerry_size_t str_end = jerry_string_to_char_buffer (str_val, str_buf, str_size);
```

Comment:
Better but there is still not 100%

+        jerry_size_t str_size = jerry_get_string_size (str_val);
+        jerry_size_t str_end = jerry_string_to_char_buffer (str_val, str_buf, str_size);

Please check the str_size before jerry_string_to_char_buffer because this might be a memory overwrite.

I also think that `jerry_size_t str_end` is unnecessary.

Reply:
You are right.